src/vs/platform/actionWidget/browser/actionWidget.ts

```diff
@@ -0,0 +1,616 @@
+		renderDisposables.add(dom.addDisposableListener(pointerBlock, dom.EventType.MOUSE_DOWN, () => pointerBlock.remove()));
+
+		// Action bar
+		let actionBarWidth = 0;
+		if (!options.fromLightbulb) {
+			const actionBar = this._createActionBar('.action-widget-action-bar', actions, options);
+			if (actionBar) {
+				widget.appendChild(actionBar.getContainer().parentElement!);
+				renderDisposables.add(actionBar);
+				actionBarWidth = actionBar.getContainer().offsetWidth;
+			}
+		}
+
+		const width = this.list.value?.layout(actionBarWidth);
+		widget.style.width = `${width}px`;
+
+		const focusTracker = renderDisposables.add(dom.trackFocus(element));
+		renderDisposables.add(focusTracker.onDidBlur(() => this.hide()));
+
+		return renderDisposables;
+	}
+
+	private _createActionBar(className: string, inputActions: ActionSet<any>, options: IActionShowOptions): ActionBar | undefined {
+		const actions = this._getActionBarActions(inputActions, options);
+		if (!actions.length) {
+			return undefined;
+		}
+
+		const container = dom.$(className);
+		const actionBar = new ActionBar(container);
+		actionBar.push(actions, { icon: false, label: true });
+		return actionBar;
+	}
+	private _getActionBarActions(actions: ActionSet<any>, options: IActionShowOptions): IAction[] {
+		const resultActions = actions.documentation.map((command): IAction => ({
+			id: command.id,
+			label: command.title,
+			tooltip: command.tooltip ?? '',
+			class: undefined,
+			enabled: true,
+			run: () => this._commandService.executeCommand(command.id, ...(command.commandArguments ?? [])),
+		}));
+
+		if (options.includeDisabledActions && actions.validActions.length > 0 && actions.allActions.length !== actions.validActions.length) {
+			resultActions.push(this.showDisabled ? {
+				id: 'hideMoreActions',
+				label: localize('hideMoreActions', 'Hide Disabled'),
+				enabled: true,
+				tooltip: '',
+				class: undefined,
+				run: () => this._toggleShowDisabled(false)
+			} : {
+				id: 'showMoreActions',
+				label: localize('showMoreActions', 'Show Disabled'),
+				enabled: true,
+				tooltip: '',
+				class: undefined,
+				run: () => this._toggleShowDisabled(true)
+			});
+		}
+		return resultActions;
+	}
+	/**
+	 * Toggles whether the disabled actions in the action widget are visible or not.
+	 */
+	private _toggleShowDisabled(newShowDisabled: boolean): void {
+		const previousCtx = this.currentShowingContext;
+
+		this.hide();
+
+		this.showDisabled = newShowDisabled;
+
+		if (previousCtx) {
+			this.show(this.list.value!, previousCtx.actions, previousCtx.anchor, previousCtx.container, previousCtx.options);
+		}
+	}
+
+	private _onWidgetClosed(didCancel?: boolean): void {
+		this.currentShowingContext = undefined;
+		this.list.value?.hide(didCancel);
+	}
+
+}
+registerSingleton(IActionWidgetService, ActionWidgetService, InstantiationType.Delayed);
+
+export abstract class ActionList<T extends IActionItem> extends Disposable implements IActionList<T> {
```

Comment:
Can this type be kept internal? Ideally the actionWidgetService should work like the `contextMenuSerive` where we pass in decorative data about the items to render, instead of requiring callers to pass in implementation details

To support this, you likely want to separate the `List` part of the code from the action data provider logic